### DISCUSSION TITLE
[Feature] Enable Linux build for mobile app

### DISCUSSION THREAD

**Discussion by luigi311** [immich-app/immich#14485] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Mobile app support for linux would be great as it will help with photo management on a linux laptop/desktop or mobile linux phone. There are more and more mobile linux phones being developed with a working camera so being able to backup those images via a familiar interface would be perfect. As of right now the only possible way to get a immich application working is to go via a android emulator but that has limitations since it wont have access to the host images. While going through the web client does allow us to view and even upload images, it does not provide a method of keeping our local images synced and would instead require us to manually selecting all the images to upload instead of the mobile app method of selecting directories and it automatically uploading missing images. Even without the background syncing ability (assuming this doesnt work) having to open the immich mobile app would be a big improvement.

Looks like flutter is used for the mobile apps and that has support for building for linux. I did see a response for a similar feature on the macos side so i figured i would create a discussion for linux itself. https://github.com/immich-app/immich/discussions/14058#discussioncomment-11206274. I was able to find this documentation on getting started with an android and desktop linux app so it might be a good starting point and hopefully it is as simple as they make it out to be.
https://docs.flutter.dev/platform-integration/linux/install-linux/install-linux-from-android

As for distribution of the linux app, it seems the most universal way with a good update system would be via flatpak which should work across all linux distributions. It would also make it so people download the application straight from you guys if you guys publish to flathub directly to avoid any issues with the source and dependencies.

I did find this discussion https://github.com/immich-app/immich/issues/1427 from 2023 but it seemed to be categorized incorrectly and the conclusion from back then was to only build a desktop sync agent. Enabling a linux build of the already built mobile app would be a better solution as it would instead keep the interaction the same across all platforms instead of having unique/limited UX for each desktop OS.

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by mmomjian** [comment]
We do not support or encourage use of the mobile app on desktop, and on platforms where it can technically be done (like you linked) it is generally broken and we do not support/bugfix for it.

A linux mobile app would be a reasonable request but I am not sure this would be a high priority anytime soon.

**Reply by luigi311** [reply]
Is there a reason to not encourage the use of the mobile app on desktop? The mobile app looks fine on desktop and would reduce the amount of work required for creating a desktop client. For mobile linux there would be no difference between it and desktop linux as they effectively are the exact same thing unlike the distinction between iOS and MacOS.

**Comment by luigi311** [comment]
Looks like getting to a starting point with linux support isnt to hard. Im not a flutter dev so this is all new to me but i was able to build and run the mobile app on my desktop linux by adding the build target

```bash
flutter create --platforms=linux .
```

and then building using

```bash
flutter build linux
```

If you get errors related to directories you need to install xdg-user-dirs.

This is what it looks like, enjoy a picture of my handsome dog
![Screenshot from 2024-12-04 18-10-39](https://github.com/user-attachments/assets/8a62dac8-73d7-4416-8ca4-e32da8873564)

Lots of work will be required to get this to a usable state though. Viewing pictures seems to work correctly. Videos and local files do not work.

**Reply by luigi311** [reply]
I was curious to try this on my mobile linux device but looks like isar errors out because it doesnt support linux_arm64 so until thats fixed upstream mobile linux on arm64, which is most devices, isnt possible.

https://github.com/isar/isar/issues/876

**Reply by luigi311** [reply]
They have a few workarounds for getting arm64 support working, editing your local isar_core.dart file to add in case Abi.linuxArm64 and then placing a libisar.so file that they compiled for arm64 in the mobile folder and that lets it actually run. Weirdly the colors seem to be off on my FLX1 but it is atleast working now and displaying with the same features as above. Heres what it looks like on my Furilabs FLX1 running phosh

![Screenshot from 2024-12-04 19-36-54](https://github.com/user-attachments/assets/7cfcfb55-8aab-435b-a246-0cb86c4ff6cd)

**Comment by luigi311** [comment]
All in all its not bad to get a base up and running. The problem is definitely going to be the plugins and making sure they support linux and adjusting all the ios/android specific fixes to make them work on linux.

**Reply by luigi311** [reply]
Some quick things
- native_video_player: Android/iOS only and doesnt seem like it will change. 
  - media_kit: Potential alternative will way more support on targets and contributors
- share_plus: Android/iOS/MacOS only for shareXfiles

**Comment by mmomjian** [comment]
Thanks for your work on this. If there are any changes on the Immich side that can make this process easier without affecting behavior/performance on Android or iOS, a PR would probably be welcome. If you want to write up your process we can probably include a link on this page if it's kept up to date and easy for someone else to use: https://immich.app/docs/community-guides

As for why we generally haven't encouraged desktop use, it's because this would be effectively another platform that we would have to test/optimize/develop for, that the team is not able to take on at this time.

**Reply by luigi311** [reply]
Gotcha ok that makes sense. Hopefully this is something that can be worked on once all the other major things have been implemented and stable has released. Mostly the mobile linux client as that would let me keep the same workflow on my mobile devices. Ill try and see what i can do though im not a flutter/dart developer so I'll see how much i can make and see if others can contribute to it before i send anything else your guys way.

**Comment by luigi311** [comment]
Did some more quick testing this is actually easier to get atleast launching and logging in on linux. The only change needed is that authentication device model and type. For now i added a fallback to unknown and submitted it in #17059 .

**Reply by luigi311** [reply]
This has been merged in. Updated the instructions in the top comments on how to get a linux build to actually compile and run.

**Comment by AndreasVerden** [comment]
@luigi311, I am very much interested in getting Immich to run on Linux mobile too. I ran into a few issues, mostly user errors but now also a missing network api.

Would you be interested in co-developing a PoC in a soft-fork?

The goal would be to demonstrate that it can be done and (hopefully) that there is actually a community of users interested in Immich on Linux mobile. Maybe this will get enough attention for @mmomjian to consider supporting the platform?

In my mind, current Linux on Mobile users should be very enthusiastic to see Immich come to their platform.

**Reply by luigi311** [reply]
i restarted tried this again since the incorrect colors should be fixed on my test phone but looks like there are some other places that need to be fixed now. I need to see what else needs to be done again to atleast get this to where it was before with it atleast being able to display my online library. 

As for getting the immich app itself to a point that its usable, its gonna require changing libraries used for certain things to make sure its using a version that supports linux. Not sure how much immich itself would be willing to make those changes to accommodate this. The immich itself is probably to big for me atleast to try to maintain a downstream fork with changes to libraries especially since im not a flutter dev.

**Reply by AndreasVerden** [reply]
I've done some digging (last week, sorry for the delay) and around one month ago two PR got merged (#25802, #25942).

They introduce a new network api with a rather interesting implementation:

https://github.com/immich-app/immich/blob/main/mobile/lib/infrastructure/repositories/network.repository.dart:
```dart
[...]
    _client?.close();
    if (Platform.isIOS) {
      final session = URLSession.fromRawPointer(clientPointer.cast());
      _client = CupertinoClient.fromSharedSession(session);
    } else {
      _client = OkHttpClient.fromJniGlobalRef(clientPointer);
    }

 [...]

 // ignore: avoid-unused-parameters
  static Future<WebSocket> createWebSocket(Uri uri, {Map<String, String>? headers, Iterable<String>? protocols}) {
    if (Platform.isIOS) {
      final session = URLSession.fromRawPointer(_clientPointer!.cast());
      return CupertinoWebSocket.connectWithSession(session, uri, protocols: protocols);
    } else {
      return OkHttpWebSocket.connectFromJniGlobalRef(_clientPointer!, uri, protocols: protocols);
    }
  }
  ``` 
  
We'd need to write a Linux/generic implementation but tbh it would probably be best to rethink how networking is done more generally... and for that we'd need the support of a maintainer (@alextran1502?)

**Reply by bo0tzz** [reply]
What sort of "rethinking" do you have in mind?

**Reply by luigi311** [reply]
The problem comes from all those very specific platform calls, are all setup in a way where it done with 

if ios do this
else do this

but that else is designed in a way that can only be used by android, which makes sense because the officially support platforms are only ios and android so if its not ios the assumption is android.

This though is the part that causes us problems as those android implementations dont work on the linux side so it needs to be designed to either use a library that abstracts this out to the different platforms correctly or 2, add in more if else conditions for the different platforms. That second method was actually the route i took for a previous issue like this that came up where for the auth headers it was setup with if ios do this and else fall back to the android calls. 
https://github.com/immich-app/immich/pull/17059

**Reply by AndreasVerden** [reply]
These are the two general options.

From what I gather, it used to be the case that option 1 was implemented until some limitation with mtls and trust stores brought that approach to its limits. So we probably don't want to go back to that.

I propose to properly[1] abstract the networking api in a way that makes it easier to extend/maintain. This has a few advantages:

1. It can be done without supporting any additional platforms
2. It would clan up existing code by simplifying and streamlining it
3. It would make it trivial to add and support additional platforms

Basically, from what I saw, the api should be an interface with implementations per platform. Only one central location should need to switch between these implementations. Open-closed.

[1]: there are some design choices in the current implementation that make it unintuitive to use which might be the source of current or future bugs. One example are unused parameters. There is also a potential security issue given that headers are/can be set globally. I haven't looked into it yet but it would be really bad if some token would be sent to third parties just because it's the default.
